proposals/0031-hlsl-vector-matrix-operations.md

```diff
@@ -156,12 +156,481 @@ vector<float, 3> ApplyNeuralMaterial(vector<half, 8> inputVector) {
+
+- `DT` - the data type of each element stored in the buffer
+
+Members:
+
+- `Buffer` - the buffer that the vector is stored in
+- `StartOffset` - the offset, in bytes, from the beginning of the buffer to
+  where the vector is located.
+
+Implementation:
+
+```c++
+namespace dx {
+namespace linalg {
+
+template <typename BUFFER, DataType DT>
+struct VectorRefImpl {
+  BUFFER Buffer;
+  uint StartOffset;
+};
+
+template <DataType DT>
+using VectorRef = VectorRefImpl<ByteAddressBuffer, DT>;
+
+template <DataType DT>
+using RWVectorRef = VectorRefImpl<RWByteAddressBuffer, DT>;
+
+} // namespace linalg
+} // namespace dx
+
+```
+
+### struct Vector
+
+> NOTE: it's possible that one resolution of [441] may remove the need for this
+> type entirely.
+
+The `Vector` struct is a wrapper around `vector`, adding an interpretation value
+that controls how the data in the vector should be interpreted. Although the
+struct can be used directly, it is likely more ergonomic to use the
+`InterpretedVector` function that's also described here.
+
+Example usage:
+
+```c++
+using namespace dx::linalg;
+
+vector<float, 128> V = 0;
+vector<float, 128> Result = Mul<float>(Matrix, InterpretedVector<DATA_TYPE_E4M3>(V));
+
+// alternative:
+Vector<float, 128, DATA_TYPE_E4M3> IV = {V};
+vector<float, 128> Result = Mul<float>(Matrix, IV);
+
+```
+
+Implementation:
+
+```c++
+namespace dx {
+namespace linalg {
+
+template <typename T, int N, DataType DT> struct Vector {
+  vector<T, N> Data;
+};
+
+template <DataType DT, typename T, int N>
+Vector<T, N, DT> InterpretedVector(vector<T, N> Vec) {
+  Vector<T, N, DT> IV = {Vec};
+  return IV;
+}
+
+} // namespace linalg
+} // namespace dx
+```
+
+[441]: https://github.com/microsoft/hlsl-specs/issues/441
+
+### Function: Mul
+
+The `Mul` function performs a matrix-vector multiplication. The matrix is stored
+in memory, while the vector comes from a variable.
+
+> TODO: add an example for packed types, and make sure they work correctly
+
+Example:
+
+```c++
+export float4 Example(float4 Input) {
+  using namespace dx::linalg;
+
+  MatrixRef<DATA_TYPE_FLOAT16, 4, 4, MATRIX_LAYOUT_MUL_OPTIMAL, true>
+      Matrix = {Buf, 0, 0};
+
+  return Mul<float>(
+      Matrix, InterpretedVector<DATA_TYPE_FLOAT16>(Input));
+}
+```
+
+Conceptual API (only the template parameters required are shown - all others are
+deduced):
+
+```c++
+namespace dx {
+namespace linalg {
+
+template<typename TYo>
+vector<TYo, M_M> Mul(MatrixRefImpl<...> Matrix, Vector<...> InputVector);
+
+} // namespace linalg
+} // namespace dx
+```
+
+See [0029] for details of this operation.
+
+> TODO: details of what this function does really belong in here as well, but
+> for now the source-of-truth is [0029].
+
+Implementation:
+
+```c++
+namespace dx {
+namespace linalg {
+
+template <typename TYo, typename TYi, int NUMi, typename M_RES, DataType IV_DT,
+          DataType M_DT, uint M_M, uint M_K, MatrixLayout M_LAYOUT,
+          bool M_TRANSPOSE>
```

Comment:
I realize this probably seems a bit nitpicky, but if I were reviewing this code itself I'd provide this feedback.
It is significantly easier to read if we have more descriptive names. Things like `OutputElTy` is more clear than `TYo`.

Reply:
This code will be things that developers see, so I think we should be reviewing it and not apologizing for it :)

I'll do a pass along these lines.

Reply:
Updated